pulp_container/app/token_verification.py

```diff
@@ -1,186 +1,119 @@
-            )
 
-    def check_authorization_token(self, authorization):
-        """
-        Check if a Bearer token is valid.
+def _contains_accessible_actions(decoded_token, repository_path, access_action):
+    """
+    Check if a client has an access permission to execute the pull/push operation.
 
-        Args:
-            authorization: A string containing a Bearer token.
+    When a client targets the root endpoint, the verifier does not necessary need to
+    check for the pull or push access permission, therefore, it is granted automatically.
 
-        Raises:
-            web.HTTPUnauthorized: A Bearer token is not valid.
+    """
+    for access in decoded_token["access"]:
+        if repository_path == access["name"]:
+            if access_action in access["actions"]:
+                return True
+            if not repository_path:
+                return True
+    return False
 
-        """
-        token = self._get_token(authorization)
-        if not self.is_token_valid(token):
-            raise web.HTTPUnauthorized(
-                headers=self._build_response_headers(),
-                reason="Access to the requested resource is not authorized. "
-                "A provided Bearer token is invalid.",
-            )
 
-    def _get_token(self, authorization):
-        """
-        Get a raw token string from the header.
+class TokenAuthentication(BaseAuthentication):
+    """
+    Token based authentication for Container Registry.
+    Clients should authenticate by passing the token key in the "Authorization"
+    HTTP header, prepended with the string "Bearer ".  For example:
+        Authorization: Bearer 401f7ac837da42b97f613d789819ff93537bee6a
+    """
 
-        This method returns a string that skips the keyword 'Bearer' and an additional
-        space from the header (e.g. "Bearer abcdef123456" -> "abcdef123456").
-        """
-        return authorization[7:]
+    keyword = "Bearer"
 
-    def _build_response_headers(self):
+    def authenticate(self, request):
         """
-        Build headers that a registry returns as a response to unauthorized access.
-
-        The method creates a value for the Www-Authenticate header. This value is used
-        by a client for requesting a Bearer token from a token server. The header
-        Docker-Distribution-API-Version is generated too to inform the client about
-        the supported schema type.
+        Check that the provided Bearer token specifies access.
         """
-        source_path = self._get_current_content_path()
-        authenticate_header = self._build_authenticate_string(source_path)
+        if settings.get("TOKEN_AUTH_DISABLED", False):
+            return (AnonymousUser(), None)
 
-        headers = {
-            "Docker-Distribution-API-Version": "registry/2.0",
-            "Www-Authenticate": authenticate_header,
-        }
-        return headers
+        try:
+            authorization_header = request.headers["Authorization"]
```

Comment:
this code changes the behaviour where we return 401 even when hitting root endpoint (/v2 )

Some registries return 401 and others 200
https://registry-1.docker.io/v2/
https://quay.io/v2/

I am ok with both of approaches however I think that returning 401 and then asking for token auth, generating the token and then verifying it in the nuance is a surplus  work because the only thing that the client cares is 1) presence of registry 2.0 header and 4) whatever code is fine except 404 because what he looks at is the header to identify whether this is a v2 api registry otherwise he needs to fall back to the v1 api.
Thoughts?

Reply:
I see both of the above examples returning 401.
Also, with the current implementation, i need to raise the `AuthenticationFailed` exception to spit out the `Www-Authenticate` header.

Reply:
ha, you are right, quay does return 401. I did through the browser so the "true" response kinda implied to me that it was a 200.